### ISSUE TITLE
IPT not giving correct occupation for large iteration

### ISSUE THREAD

**Issue by Soumeniisc** [TRIQS/triqs#416]
Dear Experts, 
I was running IPT code for bethe lattice that was given in https://triqs.ipht.cnrs.fr/1.3/tour/ipt.html. 
 i found that when i run IPT for max iteration 20 then occupation is coming to be 0.5  for all the value of U
but when i run increase the max no of iteration to 160 then i find occupation deviate from 0.5. I was expecting the it will converge better.  I added inverse FT to get g_tau and n=S.g_tau.data[9999][0][0]

for max_iteration=20
==============
U      n
0.0  0.5
0.5  0.5
1.0  0.5
1.3  0.5
1.5  0.5
1.8  0.5
2.0  0.50000000002
2.3  0.500000018426
2.5  0.500000805514
2.8  0.500069219194
3.0  0.500000005555
3.3  0.5
3.5  0.5

![hm_para_g_imag_it19_beta40 0](https://cloud.githubusercontent.com/assets/4995258/26279129/d370ae50-3dc8-11e7-9ff9-f4f7c1c691db.png)
![ipt_beta40 0_it19](https://cloud.githubusercontent.com/assets/4995258/26279196/acb1e7f0-3dca-11e7-837a-8c55842679c1.png)

for max_iteration=160
==============
U      n
0.0  0.5
0.5  0.5
1.0  0.5
1.3  0.5
1.5  0.5
1.8  0.505097863356
2.0  0.542819824766
2.3  0.529547308142
2.5  0.516734372669
2.8  0.5
3.0  0.5
3.3  0.5
3.5  0.5

![hm_para_g_imag_it159_beta40 0](https://cloud.githubusercontent.com/assets/4995258/26279138/0551e344-3dc9-11e7-89e1-e3a5e711825d.png)
![ipt_beta40 0_it159](https://cloud.githubusercontent.com/assets/4995258/26279191/a3ad8aba-3dca-11e7-95da-07300fe3a6b6.png)
Thanks in advance .

**Comment by krivenko** [comment]
Dear Soumen,

Sorry for the very late reply.
I tried to reproduce the problem, but the IPT example is completely broken (probably by 3ce348e423c76087f03117f28bdea4ccfb2c9387).

These lines
```python
        self.g0_tau << InverseFourier(self.g0_iw)
        self.sigma_tau << (self.U**2) * self.g0_tau * self.g0_tau * self.g0_tau
        self.sigma_iw << Fourier(self.sigma_tau)

        # Dyson equation to get G
        self.g_iw << inverse(inverse(self.g0_iw) - self.sigma_iw)
```
destroy tails of `self.sigma_tau`, `self.sigma_iw` and eventually of `self.g_iw`.

**Comment by mferrero** [comment]
I tried the IPT script and it should be working now. Concerning the density not being 0.5 I believe that it is because the IPT becomes numerically unstable. When you repeat the solver loop many times you eventually propagate a bit of numerical noise that destroys the expected symmetry of the Green's function. You can see that on your plot of the spectral function. I guess one should enforce symmetries by hand if you want to increase the number of iterations.

**Comment by Soumeniisc** [comment]
Hi mferrero,

Sorry for coming back to this after long time. You mentioned "I tried the IPT script and it should be working now" . Do you remember what was the fix that was maid in the IPT script? 

I am worried because u also mentioned one should enforce symmetries. But in the current script it's not shown. Am i missing something?

Thanks

**Comment by mferrero** [comment]
Hi,

We have been working on the library and at some point we had some issues with the high frequency tails of the Green functions. This is why for some time the IPT example did not work. But then we fixed the library and when I tested the example, it was working fine. It should still be working today. In any case, I did not make any changes in the IPT script, I just used the latest version of the library.

In the example, we do not try to enforce particle-hole symmetry, which should be present in the problem at half-filling. It may be that if the number of iterations is large, numerical noise propagates and eventually this symmetry is broken. A way to enforce it manually is to set the real part of the Green function to zero.
